### PR TITLE
Allow recurrent contacts

### PR DESCRIPTION
**Pull request by janosg**
### Current behavior

No recurrent contacts are possible

### Desired behavior

Recurrent contacts in households or school classes can be modeled. 

### Solution / Implementation

When setting the "model" entry in a contact model to "meet_group", all people in the assort_by group of that contact model meet in

### CODE REVIEW

File: sid/simulate.py [hunk 1]
```diff
@@ -116,14 +116,70 @@ def simulate(
...
+        contact_models (dict): see :ref:`contact_models`
+
+    Returns:
+        assort_bys (dict): Keys are names of contact models, values are lists with the
+            assortative_by variables of the model.
+
+    """
+    assort_bys = {}
+    for model_name, model in contact_models.items():
+        assort_by = model.get("assortative_by", None)
```
[the remark this answers is not in the document]
Comment (janosg) [reply sid/simulate.py:155]: Yes, I wanted to move to assortative by, but was too used to assort_by already, so my switch was very incomplete. Will move back to assort_by.

File: sid/simulate.py [hunk 2]
```diff
@@ -116,14 +116,70 @@ def simulate(
...
+            )
+            assort_by = []
+        elif not assort_by:
+            assort_by = []
+        elif isinstance(assort_by, str):
+            assort_by = [assort_by]
+        elif isinstance(assort_by, list):
+            pass
+        else:
+            raise ValueError()
```
[the remark this answers is not in the document]
Comment (janosg) [reply sid/simulate.py:169]: Changed it manually.

File: sid/tests/test_contacts.py [hunk 3]
```diff
@@ -126,20 +134,74 @@ def _sample_data_for_calculate_infections_numba(
...
+        group_codes.reshape(-1, 1),
+        group_probs_list,
+        indexers_list,
         infection_prob,
+        is_meet_group,
+        loop_order,
+    )
+
+
+# test calculate_infections with only one recurrent contact model and very few states
```
[the remark this answers is not in the document]
Comment (janosg) [reply sid/tests/test_contacts.py:168]: Thanks.

File: sid/tests/test_contacts.py [hunk 4]
```diff
@@ -126,20 +134,74 @@ def _sample_data_for_calculate_infections_numba(
...
     for group in range(n_groups):
         indexer.append(np.where(group_codes == group)[0])
 
+    indexers_list = NumbaList()
+    indexers_list.append(indexer)
+
     if infection_prob is None:
-        infection_prob = np.random.uniform()
+        ip = np.random.uniform()
+        infection_prob = np.array([ip])
```
[the remark this answers is not in the document]
Comment (janosg) [reply sid/tests/test_contacts.py:149]: This was actually your code. But I think for tests we can leave it like this.

File: sid/contacts.py [hunk 5]
```diff
@@ -68,32 +72,53 @@ def calculate_infections(states, contacts, params, indexer, group_probs, seed):
         states (pandas.DataFrame): Copy of states with updated immune column.
 
     """
+    is_meet_group = np.array([k not in group_probs for k in indexers])
```
[the remark this answers is not in the document]
Comment (janosg) [reply sid/contacts.py:75]: I find all of them not ideal, but will leave it for now.

File: sid/simulate.py [hunk 6]
```diff
@@ -116,14 +116,70 @@ def simulate(
...
+        assort_bys (dict): Keys are names of contact models, values are lists with the
+            assortative_by variables of the model.
+
+    """
+    assort_bys = {}
+    for model_name, model in contact_models.items():
+        assort_by = model.get("assortative_by", None)
+        if assort_by is None:
+            warnings.warn(
+                "Not specifying 'assortative_by' significantly raises runtime. "
```
[the remark this answers is not in the document]
Comment (janosg) [reply sid/simulate.py:158]: Done.

File: sid/simulate.py [hunk 7]
```diff
@@ -78,20 +71,27 @@ def simulate(
...
-    indexer = create_group_indexer(states, assort_by)
-    first_probs = create_group_transition_probs(states, assort_by, params)
+    indexers = {}
+    first_probs = {}
+    for model_name, assort_by in assort_bys.items():
+        indexers[model_name] = create_group_indexer(states, assort_by)
+        if contact_models[model_name]["model"] != "meet_group":
+            first_probs[model_name] = create_group_transition_probs(
+                states, assort_by, params
+            )
```
[the remark this answers is not in the document]
Comment (janosg) [reply sid/simulate.py]: Done, thanks.

File: sid/tests/test_contacts.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tobiasraabe) [reply sid/tests/test_contacts.py:149]: Ah, missed that it is the sampling function. Yes, then it is fine.

**Pull request merged**